pkg/services/sqlstore/migrations/ualert/migration_test.go

```diff
@@ -0,0 +1,601 @@
+		Settings:     settings,
+		NewStateDate: now,
+		Created:      now,
+		Updated:      now,
+	}
+}
+
+// createDashboard creates a dashboard for inserting into the test database.
+func createDashboard(t *testing.T, id int64, orgId int64, uid string) *models.Dashboard {
+	t.Helper()
+	return &models.Dashboard{
+		Id:      id,
+		OrgId:   orgId,
+		Uid:     uid,
+		Created: now,
+		Updated: now,
+		Title:   uid, // Not tested, needed to satisfy contraint.
+	}
+}
+
+// createDatasource creates a ddatasource for inserting into the test database.
+func createDatasource(t *testing.T, id int64, orgId int64, uid string) *models.DataSource {
+	t.Helper()
+	return &models.DataSource{
+		Id:      id,
+		OrgId:   orgId,
+		Uid:     uid,
+		Created: now,
+		Updated: now,
+		Name:    uid, // Not tested, needed to satisfy contraint.
+	}
+}
+
+// Clean input tables for each test case.
+func teardown(t *testing.T, x *xorm.Engine) {
+	_, err := x.Exec("DELETE from alert")
+	require.NoError(t, err)
+	_, err = x.Exec("DELETE from alert_notification")
+	require.NoError(t, err)
+	_, err = x.Exec("DELETE from dashboard")
+	require.NoError(t, err)
+	_, err = x.Exec("DELETE from data_source")
+	require.NoError(t, err)
+}
+
+// Setup and insert data into legacy alerting tables needed for migration.
+func setupLegacyAlertsTables(t *testing.T, x *xorm.Engine, legacyChannels []*models.AlertNotification, alerts []*models.Alert) {
+	t.Helper()
+
+	// Setup dashboards.
+	dashboards := []models.Dashboard{
+		*createDashboard(t, 1, 1, "dash1-1"),
+		*createDashboard(t, 2, 1, "dash2-1"),
+		*createDashboard(t, 3, 2, "dash3-2"),
+		*createDashboard(t, 4, 2, "dash4-2"),
+	}
+	_, errDashboards := x.Insert(dashboards)
+	require.NoError(t, errDashboards)
+
+	// Setup data_sources.
+	dataSources := []models.DataSource{
+		*createDatasource(t, 1, 1, "ds1-1"),
+		*createDatasource(t, 2, 1, "ds2-1"),
+		*createDatasource(t, 3, 2, "ds3-2"),
+		*createDatasource(t, 4, 2, "ds4-2"),
+	}
+	_, errDataSourcess := x.Insert(dataSources)
+	require.NoError(t, errDataSourcess)
+
+	if len(legacyChannels) > 0 {
+		_, channelErr := x.Insert(legacyChannels)
+		require.NoError(t, channelErr)
+	}
+
+	if len(alerts) > 0 {
+		_, alertErr := x.Insert(alerts)
+		require.NoError(t, alertErr)
+	}
+}
+
+type postableUserConfig struct {
+	AlertmanagerConfig postableApiAlertingConfig `yaml:"alertmanager_config" json:"alertmanager_config" xorm:"alertmanager_config"`
+}
+
+type postableApiAlertingConfig struct {
+	Route     *route                        `yaml:"route,omitempty" json:"route,omitempty"`
+	Receivers []*ualert.PostableApiReceiver `yaml:"receivers,omitempty" json:"receivers,omitempty"`
+}
+
+type route struct {
+	Receiver string   `yaml:"receiver,omitempty" json:"receiver,omitempty"`
+	Matchers []string `yaml:"matchers,omitempty" json:"matchers,omitempty"`
+	Routes   []*route `yaml:"routes,omitempty" json:"routes,omitempty"`
+	name     string   // Used to help attach Matchers to expected results.
+}
```

Comment:
I know we discussed this offline, but making the comment for posterity - I'm not convinced this is the right approach.

I would expect the serialized Alertmanager configuration to match whatever we use in the `nglaert` package as opposed to custom structs for assertion. I think this has the potential to go out of sync and give us a false sense of confidence in these tests unless we're explicitly testing the end result.

That being said, these tests are great and are def an improvement so I'm happy with the outcome here.

Reply:
You're right that these test-only snapshots are not great. I mostly had issue with unmarshalling using `config.go:Load(...)` for a couple reasons, one being that the `ualert` package itself uses a snapshot instead of the `ngalert` definitions. I've removed them now and am unmarshalling directly to the `ualert` package snapshots instead of these redundant ones.